subworkflows/nf-core/fasta_explore_search_plot_tidk/main.nf

```diff
@@ -0,0 +1,86 @@
+include { SEQKIT_SEQ as FILTER_BY_LENGTH            } from '../../../modules/nf-core/seqkit/seq'
+include { SEQKIT_SORT as SORT_BY_LENGTH             } from '../../../modules/nf-core/seqkit/sort'
+include { TIDK_EXPLORE                              } from '../../../modules/nf-core/tidk/explore'
+include { TIDK_SEARCH as TIDK_SEARCH_APRIORI        } from '../../../modules/nf-core/tidk/search'
+include { TIDK_SEARCH as TIDK_SEARCH_APOSTERIORI    } from '../../../modules/nf-core/tidk/search'
+include { TIDK_PLOT as TIDK_PLOT_APRIORI            } from '../../../modules/nf-core/tidk/plot'
+include { TIDK_PLOT as TIDK_PLOT_APOSTERIORI        } from '../../../modules/nf-core/tidk/plot'
+
+
+workflow FASTA_EXPLORE_SEARCH_PLOT_TIDK {
+
+    take:
+    ch_fasta                // channel: [ val(meta), [ fasta ] ]
+    ch_apriori_sequence     // channel: [ val(meta), val(sequence) ]; Optional: Set to [] if not needed
+
+    main:
+    ch_versions = Channel.empty()
+
+    // MODULE: SEQKIT_SEQ as FILTER_BY_LENGTH
+    FILTER_BY_LENGTH ( ch_fasta )
+
+    ch_filtered_fasta       = FILTER_BY_LENGTH.out.fastx
+    ch_versions             = ch_versions.mix(FILTER_BY_LENGTH.out.versions.first())
+
+    // MODULE: SEQKIT_SORT as SORT_BY_LENGTH
+    SORT_BY_LENGTH ( ch_filtered_fasta )
+
+    ch_sorted_fasta         = SORT_BY_LENGTH.out.fastx
+    ch_versions             = ch_versions.mix(SORT_BY_LENGTH.out.versions.first())
+
+    // TIDK_EXPLORE
+    TIDK_EXPLORE ( ch_filtered_fasta )
+
+    ch_top_sequence         = TIDK_EXPLORE.out.top_sequence
+    ch_versions             = ch_versions.mix(TIDK_EXPLORE.out.versions.first())
+
+    // TIDK_SEARCH as TIDK_SEARCH_APRIORI
+    ch_apriori_inputs       = ch_sorted_fasta
+                            | join(
```

Comment:
Most of the use cases can be covered by making ch_apriori_sequence a value channel such as val(sequence) and then applying combine at line 39, as a species-level _a priori_ sequence can work for multiple haplotypes from the same species. AssemblyQC, for example, suggests [kingdom level sequences](https://github.com/Plant-Food-Research-Open/assembly_qc/blob/0c26eecbb7be7f64ab9b7049a34d371a44faf9d1/nextflow.config#L101). 

In writing this sub-workflow I defaulted to the most general case so that it can handle haplotypes from different species with different _a priori_ sequences.

I agree with your concern. I have seen unexpected resume behaviour from NextFlow when the meta map contains child maps or lists. Here, all we need to join on is the `id` key. I'll make that change.

Reply:
So will `ch_sorted_fasta` only ever contain one fasta, otherwise the `join` will discard the rest after the first match?

Reply:
That's a good point. Let's think through this.

Let's say the workflow is asked to process `N` fasta. In that case, both `ch_fasta` and `ch_apriori_sequence` should emit `N` values. If we use the example code from nf-test:

```groovy
ch_fasta_input = Channel.of(
    [
        [ id:'test' ],
        file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
    ],
    [
        [ id:'test2' ],
        file(params.test_data['homo_sapiens']['genome']['genome_21_fasta'], checkIfExists: true)
    ]
)
ch_seq_input = Channel.of(
    [
        [ id:'test' ],
        'TTAGGG'
    ],
    [
        [ id:'test2' ],
        'TTAGGG'
    ]
)
```

The cardinality and `meta` map of ch_fasta will be propagated to `ch_sorted_fasta` through `ch_filtered_fasta`. In this case, the `join` will not break.

If the user does not provide `ch_seq_input` for one of the fasta in `ch_fasta_input `, the workflow will still work but only for the fasta which has a corresponding sequence in `ch_seq_input `. This case is tested here: https://github.com/nf-core/modules/blob/510f55602fdec308d61049394da78efacc687231/subworkflows/nf-core/fasta_explore_search_plot_tidk/tests/main.nf.test#L55

Reply:
Here is another option. Although, it might be confusing for some users and I don't think it is supported by nf-core/tools schema linter.

```groovy
ch_apriori_sequence     // channel: [ val(meta), val(sequence) ]
                        // Or val(sequence) which will be applied to each fasta in ch_fasta
                        // Optional: Set to [] if not needed

// and then...

ch_apriori_inputs       = ch_apriori_sequence instanceof Sring
                        ? ( ch_sorted_fasta.combine(ch_apriori_sequence) )
                        : (
                            ch_sorted_fasta
                            | map { meta, fasta -> [ meta.id, meta, fasta ] }
                            | join(
                                ( ch_apriori_sequence ?: Channel.empty() )
                                | map { meta, seq -> [ meta.id, seq ] }
                            )
                            | map { id, meta, fasta, seq -> [ meta, fasta, seq ] }
                        )
```

Reply:
The first option is fine. I'm just checking how this intended to be used as it's not always clear, and subworkflows don't act in the same way as processes. A subworkflow is basically just a zip-tie for channels, so one needs to make sure that the inputs are intended that way.